### ISSUE TITLE
FATAL Error with Composer classmap-authoritative

### ISSUE THREAD

**Issue by skyzyx** [Spomky-Labs/jose#291]
```
PHP Fatal error:  Trait 'Jose\Object\JWKSetPEM' not found in /var/www/vendor/spomky-labs/jose/src/Object/JWKSet.php on line 20
```

This happens when I use `composer dump-autoload -oa`, but not when I run `composer dump-autoload -o`. Something about how the classmap is being generated.

This is a recent issue. I've been using `composer dump-autoload -oa` with this project for a while, and I only started receiving these errors recently. Not sure if the problem is here, or with Composer, but this is the only dependency that I'm seeing errors with.

----

This code isn't runnable as I'm posting it here, but it should represent the JOSE-related code that is relevant.

```php
use Jose\Checker\AudienceChecker;
use Jose\Checker\CheckerManager;
use Jose\Checker\CriticalHeaderChecker;
use Jose\Checker\ExpirationTimeChecker;
use Jose\Checker\IssuedAtChecker;
use Jose\Checker\NotBeforeChecker;
use Jose\Factory\JWKFactory;
use Jose\Loader;

// @codingStandardsIgnoreStart
if (!(
    $request->hasHeader('Authorization')
    && \preg_match('/^Bearer (.*)$/', $request->getHeader('Authorization')[0], $m)
)) {
    throw new Exception('A bearer token is required.');
}
// @codingStandardsIgnoreEnd

$this->container['jwt'] = $m[1];

$sigIndex = 0;
$loader   = new Loader();
$jwk      = JWKFactory::createFromKeyFile(APP_ROOT . '/keys/internal-service.pub', null, [
    'use' => 'sig',
    'alg' => 'RS256',
]);

// Validate the JWT
$this->container['jws'] = function (Container $c) use ($loader, $jwk, $sigIndex) {
    return $loader->loadAndVerifySignatureUsingKey(
        $c['jwt'],
        $jwk,
        ['RS256'],
        $sigIndex
    );
};

// Need to do validation
$checker_manager = new CheckerManager();

// We add the claims checkers.
$checker_manager->addClaimChecker(new ExpirationTimeChecker());
$checker_manager->addClaimChecker(new IssuedAtChecker());
$checker_manager->addClaimChecker(new NotBeforeChecker());
$checker_manager->addClaimChecker(new AudienceChecker('internal-service'));

// We add the header checker (crit).
$checker_manager->addHeaderChecker(new CriticalHeaderChecker());

// We verify the signature at index $signature_index of the JWS $jws.
// If the verification failed, an exception will be thrown.
$checker_manager->checkJWS($this->container['jws'], $sigIndex);

// Store the current user identifier
$this->container['jwt_uuid'] = function (Container $c) use ($sigIndex) {
    return $c['jws']->getClaim('sub');
};

// Store the current user email
$this->container['jwt_email'] = function (Container $c) use ($sigIndex) {
    return $c['jws']->getClaim('email');
};
```

**Comment by Spomky** [comment]
Thank you for that detailed report.
I will try to reproduce that bug.

**Comment by skyzyx** [comment]
So, I think I've narrowed-down what's happening.

I develop inside of a CentOS 7.4 VM running on top of macOS 10.13. I've configured NFS between the two file systems. I've been using this Vagrant-based setup for over 3 years without issue.

However, with the move to the latest macOS, I adopted the new Apple Filesystem (APFS). What I seem to be observing is that when I run `composer install -oa` from my CentOS VM, I run into this issue. However, when I run it from my macOS environment, I don't. (I think. I'm still confirming.)

I realized I was observing more missing classes from the classmap earlier today, and I've also run into issues with tools like PHP CS Fixer and PHP CodeSniffer complaining about duplicate classes — when really, it's just getting duplicate results from the NFS mount.

So it doesn't appear to be an issue with your package, or even with Composer — but rather, with running an NFS mount on top of APFS. :/ Feel free to resolve this issue. I'll reach out to Apple, Vagrant, VMware, and the NFS dev teams and file tickets with them.
